### PR TITLE
Jupyter notebook graphics compression

### PR DESCRIPTION
**Pull request by hshe824**
This patch affords trimming of redundant Jupyter notebook information contained within TCanvas .JSON objects. Specifically this includes the removing the list of TColors which reduces the size of a benchmark notebook (a compiled test notebook of Root Primer graphics) by up to 80% upon preliminary tests.

### CODE REVIEW

**Comment by martinmine** [comment]
@phsft-bot build!

**Comment by dpiparo** [comment]
Would it be possible to test this with Python3 as well?

File: bindings/pyroot/JupyROOT/utils.py [hunk 1]
```diff
@@ -390,14 +394,51 @@ def _canJsDisplay(self):
...
+	{'_typename': 'TCanvas', 'fPrimitives': {'arr': [{'arr': [{'_typename': 'NotTColor'}], '_typename': 'NotObjArray', 'name': 'ListOfColors'}]}}
+	"""
+	
+	#Only TCanvas JSON objects have TColors to remove in the first place
+	if "_typename" in object_json and object_json["_typename"]!="TCanvas":
+	    return object_json
+
+	#TColor objects are known to have this fixed nested structure in the TCanvas as follows:
+	#TCanvas -> fPrimitives-> arr-> TObjArray (an element in arr)
+	if "fPrimitives" in object_json:
```
Comment (hshe824) [inline bindings/pyroot/JupyROOT/utils.py]: done on lines 419-420

**Comment by dpiparo** [comment]
@phsft-bot build!

File: bindings/pyroot/JupyROOT/utils.py [hunk 2]
```diff
@@ -390,14 +393,62 @@ def _canJsDisplay(self):
                     return False
         return True
 
-
+    def _removeTColors(self,object_json):
+	'''
+	This function trims the json by trimming the list of TColors that are not used in JS rendering, and hence can
+	be safety removed, reducing the size of notebooks with TCanvases markedly
```
Comment (dpiparo) [inline bindings/pyroot/JupyROOT/utils.py]: safety -> safely?

**Comment by linev** [comment]
Hi,

But this only works when default palette is used.
And even default ROOT palette may differ from default JSROOT palette.

If user changes color palette, list of colors will be different. 
Do you check this?

**Comment by linev** [comment]
@hshe824, why you want to add ConvertToJSONGraphics to TBufferJSON class?
Your current method does not use any internal functions of TBufferJSON class

**Comment by hshe824** [comment]
Hi @linev,

I am not familiar with how the palette works in ROOT and JSROOT, as I am still getting started with this. What would your advice be on how to best check and test this?

For one of my tasks, I am currently in the process of implementing the same functionality of trimming (TColors) in C++ rather than python as requested by Danilo. I was going to do this in the ConvertToJSONGraphics method, which would replace the respective python methods.

Cheers,
Harry

**Comment by linev** [comment]
> What would your advice be on how to best check and test this?

I do not have good solution for you.
As I mention - even when default palette is used in ROOT, there is no guarantee that default palette is exactly the same as default palette in JSROOT. 

Probably, one could create special C++ class, which represent all palette colors in more compact way - list of TColor objects has a lot of unused information.

> I was going to do this in the ConvertToJSONGraphics method, which would replace the respective python methods.

My argument - TBufferJSON is not a proper place for such function. You should put it in separate sources, related to iPython graphics

**Comment by hshe824** [comment]
Thank you @linev, what location do you suggest I put this function? I am not familiar with the graphics location structure. Cheers

**Comment by dpiparo** [comment]
I think we need a piece of code in the TPad, as discussed with @couet , to determine if the colourse can be skipped or not, something that tells us if the user went for custom colours. With that piece in place, we will selectively trim the json.

**Pull request closed without merging**